README.md
## DaedalusBLN Kodi Repository

This unofficial Kodi add-ons repository is based on [drinfernoo](https://github.com/drinfernoo)'s **repository.example** template and the [Add-on zips/Repo XML Generator](https://github.com/drinfernoo/repository.example/blob/master/_repo_xml_generator.py) python script.

To automate the process of updating/adding to the repo even further and eliminate the continuous need of access to a personal computer, I have setup a continuous integration configuration script for [Travis-CI](https://travis-ci.com) that runs the xml/zip generator python script on a virtual machine in the cloud each time I push a commit of a new or updated unzipped add-on folder to the root directory of this repository and makes and pushes another followup commit with the generated/updated contents of the _zips_ folder.

Feel free to look around or open an issue to ask me how you can do it too, it's simpler than you think.

Have fun with the add-ons!

_Daedalus_


skin.grid/changelog.txt
2.5.2
- Added Forced Views (Initial commit)

2.5.1
- Improved listing of HDR in information dialog and list view, while still preserving correct media source

2.5
- Improved Leia compatibility
- Improved recognition of WEB-DL files when determining media source
- Added watch-count to media information dialog
- Removed unused views and "more" option from information dialog
- Fixed duration being listed incorrectly in slide view
- Added tagline to description in slide view
- Changed some strings from skin-specific to native Kodi strings.
- Fixed missing resume-indicator for items that were unwatched

2.0
- First official Leia release
- Updated default font to latest version (Cyrillic support)
- Tweaks to volumebar behaviour
- Fixed some localisation related issues
- Various small fixes and tweaks

1.3
- Option to disable media info on pause
- List view can now be set to portrait or landscape on a per instance basis
- PVR artwork can be disabled
- First run setup
- Landscape list view available to seasons
- Subtitle search dialog shows number of subtitles
- Various options to display playback time on seekbar
- Improved indication of recording timers in TV Guide
- Fallback icons for channel logo's
- Fixed double dislay of stream info for radio
- Alternate, capitalized fontset
- Finish time available in video info window
- Extra default shortcuts for stupid entitled people
- Channel manager refinement
- Switch channel groups in OSD Channel Guide using directional buttons
- Lots of PVR tweaks and fixes (thanks joostzilla & h.udo)
- Various fixes, optimizations and tweaks

1.2.48
- First official release

1.0.0
- everything


script.autoruns/default.py
# -*- coding: utf-8 -*-

""" Autoruns
    2015 fightnight"""

import xbmc,xbmcaddon,xbmcgui,xbmcplugin,urllib,os,re,sys

def list_addons():
      #info directory
      addDir('[COLOR blue][B]%s[/B][/COLOR]' % (translate(30001)),'None',None,os.path.join(xbmc.translatePath(xbmcaddon.Addon().getAddonInfo('path')).decode('utf-8'),'icon.png'))

      #get the path of addons
      pathofaddons = xbmc.translatePath('special://home/addons').decode('utf-8')

      #list with addons
      listofaddons = os.listdir(pathofaddons)
      for individual_addon in listofaddons:
            #path to individual addon, cycle for all the addons
            path_to_addon = os.path.join(pathofaddons, individual_addon)

            #define addon.xml path
            addon_xml_path=os.path.join(path_to_addon,'addon.xml')

            #check the existence of addon.xml, if true, we continue
            if os.path.exists(addon_xml_path):

                  #get addon.xml content
                  xml_content=openfile(addon_xml_path)

                  if re.search('point="xbmc.service"',xml_content):
                        #addon with service on
                        addDir('%s (on)' % (individual_addon),path_to_addon,1,os.path.join(path_to_addon,'icon.png'))
                  elif re.search('point="xbmc.pass"',xml_content):
                        #addon with service off
                        addDir('[B][COLOR gold]%s[/B] (off)[/COLOR]' % (individual_addon),path_to_addon,1,os.path.join(path_to_addon,'icon.png'))
                  else:
                        #addon with no service
                        pass

def change_state(name,path):
      #define addon.xml path to change
      addon_xml_path=os.path.join(path,'addon.xml')

      #get addon.xml content
      content=openfile(addon_xml_path)
      
      if re.search('COLOR gold',name):
            #service off to on, so we change from fake variable to service variable
            content=content.replace('point="xbmc.pass"','point="xbmc.service"')
      else:
            #service on to off, so we change from service variable to fake variable
            content=content.replace('point="xbmc.service"','point="xbmc.pass"')

      #change state on addon.xml
      savefile(addon_xml_path,content)

      #refresh the list
      xbmc.executebuiltin("Container.Refresh")
      

def openfile(path_to_the_file):
    try:
        fh = open(path_to_the_file, 'rb')
        contents=fh.read()
        fh.close()
        return contents
    except:
        print "Wont open: %s" % filename
        return None

def savefile(path_to_the_file,content):
    try:
        fh = open(path_to_the_file, 'wb')
        fh.write(content)  
        fh.close()
    except: print "Wont save: %s" % filename

def addDir(name,path,mode,iconimage):
      return xbmcplugin.addDirectoryItem(handle=int(sys.argv[1]),url="%s?path=%s&mode=%s&name=%s" % (sys.argv[0],urllib.quote_plus(path),mode,urllib.quote_plus(name)),listitem=xbmcgui.ListItem(name,iconImage="DefaultFolder.png", thumbnailImage=iconimage),isFolder=False)

def get_params():
      param=[]
      paramstring=sys.argv[2]
      if len(paramstring)>=2:
            params=sys.argv[2]
            cleanedparams=params.replace('?','')
            if (params[len(params)-1]=='/'):
                  params=params[0:len(params)-2]
            pairsofparams=cleanedparams.split('&')
            param={}
            for i in range(len(pairsofparams)):
                  splitparams={}
                  splitparams=pairsofparams[i].split('=')
                  if (len(splitparams))==2:
                        param[splitparams[0]]=splitparams[1]                 
      return param

def translate(text):
      return xbmcaddon.Addon().getLocalizedString(text).encode('utf-8')

params=get_params()
path=None
name=None
mode=None

try: path=urllib.unquote_plus(params["path"])
except: pass
try: name=urllib.unquote_plus(params["name"])
except: pass
try: mode=int(params["mode"])
except: pass

if mode==None: list_addons()
elif mode==1: change_state(name,path)
                       
xbmcplugin.endOfDirectory(int(sys.argv[1]))


plugin.audio.soundcloud/resources/lib/content/client.py
__author__ = 'bromix'

from resources.lib import nightcrawler
from resources.lib.nightcrawler.exception import ProviderException
from resources.lib.content import items


class Client(nightcrawler.HttpClient):
    CLIENT_ID = '40ccfee680a844780a41fbe23ea89934'
    CLIENT_SECRET = '26a5240f7ee0ee2d4fa9956ed80616c2'

    def __init__(self, access_token='', client_id='', client_secret='', items_per_page=50):
        nightcrawler.HttpClient.__init__(self, default_header={'Accept-Encoding': 'gzip',
                                                               'Host': 'api.soundcloud.com:443',
                                                               'Connection': 'Keep-Alive',
                                                               'User-Agent': 'SoundCloud-Android/15.09.14-release (Android 5.0.1; samsung GT-I9505)'})
        self._access_token = access_token
        self._items_per_page = items_per_page

        # set client id with fallback
        self._client_id = self.CLIENT_ID
        if client_id:
            self._client_id = client_id
            pass

        # set client secret with fallback
        self._client_secret = self.CLIENT_SECRET
        if client_secret:
            self._client_secret = client_secret
            pass
        pass

    def get_access_token(self):
        return self._access_token

    def _create_url(self, path, user_id=None):
        if user_id:
            if user_id == 'me':
                path = 'me/%s' % path.strip('/')
                pass
            else:
                path = 'users/%s/%s' % (user_id, path.strip('/'))
                pass
            pass

        return 'https://api.soundcloud.com:443/%s' % path

    def _create_params(self, page=1):
        params = {'limit': str(self._items_per_page)}
        if page > 1:
            params['offset'] = str((page - 1) * self._items_per_page)
        return params

    def _handle_error(self, response):
        status_class = response.status_code/100
        # accept 2XX and 3XX
        if status_class != 2 and status_class != 3 and status_class != 5:
            json_data = response.json()
            error_message = 'HTTP error %d' % response.status_code
            if 'error' in json_data:
                error_message = json_data['error']
                pass

            if 'errors' in json_data:
                errors = json_data['errors']
                if len(errors) > 0:
                    error_message = errors[0].get('error_message', '')
                pass

            # last fallback
            if not error_message:
                error_message = response.headers.get('status', error_message)
                pass
            raise nightcrawler.CredentialsException(error_message)
        pass

    def resolve_url(self, url):
        params = {'url': url}
        response = self._request(self._create_url('resolve'),
                                 headers={'Accept': 'application/json'},
                                 params=params)
        self._handle_error(response)
        return items.convert_to_item(response.json())

    def get_user(self, user_id):
        # self.update_access_token()
        response = self._request(self._create_url('', user_id=user_id),
                                 headers={'Accept': 'application/json'})
        self._handle_error(response)
        return items.convert_to_item(response.json())

    def get_track(self, track_id):
        response = self._request(self._create_url('tracks/%s' % unicode(track_id)),
                                 headers={'Accept': 'application/json'})
        self._handle_error(response)
        return items.convert_to_item(response.json())

    def get_track_url(self, track_id):
        response = self._request(self._create_url('tracks/%s/stream' % unicode(track_id)),
                                 headers={'Accept': 'application/json'}, allow_redirects=False)
        self._handle_error(response)
        if response.status_code == 302:
            return response.headers['location']

        return response.url

    def get_trending(self, category='music', page=1):
        if not category.lower() in ['music', 'audio']:
            raise ProviderException('Unknown category "%s"' % category)

        params = self._create_params(page)
        response = self._request(self._create_url('app/mobileapps/suggestions/tracks/popular/%s' % category),
                                 headers={'Accept': 'application/json'},
                                 params=params)
        self._handle_error(response)
        return items.convert_to_items(response.json(), mobile_conversion=True)

    def get_genre(self, genre, page=1):
        params = self._create_params(page)
        response = self._request(self._create_url('app/mobileapps/suggestions/tracks/categories/%s' % genre),
                                 headers={'Accept': 'application/json'},
                                 params=params)
        self._handle_error(response)
        return items.convert_to_items(response.json(), mobile_conversion=True)

    def get_categories(self):
        def _process_category(_category, _result, _json_data):
            for _item in _json_data:
                _result.append({'type': 'folder',
                                'title': _item['title']})
                pass
            pass

        response = self._request(self._create_url('app/mobileapps/suggestions/tracks/categories'),
                                 headers={'Accept': 'application/json'})
        self._handle_error(response)

        json_data = response.json()
        result = {'audio': [], 'music': []}
        for key in result:
            _process_category(key, result[key], json_data.get(key, []))
            pass

        return result

    def get_recommended_for_track(self, track_id, page=1):
        params = self._create_params(page)
        params['linked_partitioning'] = '1'
        response = self._request(self._create_url('tracks/%s/related' % str(track_id)),
                                 headers={'Accept': 'application/json'},
                                 params=params)
        self._handle_error(response)
        return items.convert_to_items(response.json())

    def get_tracks(self, user_id, page=1):
        params = self._create_params(page)
        params['linked_partitioning'] = '1'
        response = self._request(self._create_url('tracks', user_id=user_id),
                                 headers={'Accept': 'application/json'},
                                 params=params)
        self._handle_error(response)
        return items.convert_to_items(response.json())

    def get_favorites(self, user_id, page=1):
        params = self._create_params(page)
        params['linked_partitioning'] = '1'
        response = self._request(self._create_url('favorites', user_id=user_id),
                                 headers={'Accept': 'application/json'},
                                 params=params)
        self._handle_error(response)
        return items.convert_to_items(response.json())

    def get_following(self, user_id, page=1):
        params = self._create_params(page)
        params['linked_partitioning'] = '1'
        response = self._request(self._create_url('followings', user_id=user_id),
                                 headers={'Accept': 'application/json'},
                                 params=params)
        self._handle_error(response)
        return items.convert_to_items(response.json())

    def get_follower(self, user_id, page=1):
        params = self._create_params(page)
        params['linked_partitioning'] = '1'
        response = self._request(self._create_url('followers', user_id=user_id),
                                 headers={'Accept': 'application/json'},
                                 params=params)
        self._handle_error(response)
        return items.convert_to_items(response.json())

    def get_playlist(self, playlist_id):
        response = self._request(self._create_url('playlists/%s' % unicode(playlist_id)),
                                 headers={'Accept': 'application/json'})
        self._handle_error(response)
        json_result = response.json()

        # we transform the result of the playlist, so we can use the common method
        json_result['collection'] = json_result['tracks']
        return items.convert_to_items(json_result, process_tracks_of_playlist=True)

    def get_playlists(self, user_id, page=1):
        params = self._create_params(page)
        params['linked_partitioning'] = '1'
        response = self._request(self._create_url('playlists', user_id=user_id),
                                 headers={'Accept': 'application/json'},
                                 params=params)
        self._handle_error(response)
        return items.convert_to_items(response.json())

    def get_likes(self, user_id, page=1):
        params = self._create_params(page)
        params['linked_partitioning'] = '1'
        if page > 1:
            params['page_size'] = params['offset']
            params['page_number'] = unicode(page)
            pass

        response = self._request(self._create_url('e1/users/%s/likes' % unicode(user_id)),
                                 headers={'Accept': 'application/json'},
                                 params=params)
        self._handle_error(response)
        return items.convert_to_items(response.json())

    def search(self, search_text, category='tracks', page=1):
        if not category in ['tracks', 'users', 'playlists']:
            raise ProviderException('Unknown category "%s"' % category)

        params = {'limit': str(self._items_per_page),
                  'q': search_text,
                  'linked_partitioning':page}
        # if page > 1:
        #     params['offset'] = str((page - 1) * self._items_per_page)

        response = self._request(self._create_url('%s/' % category),
                                 headers={'Accept': 'application/json'},
                                 params=params)
        self._handle_error(response)
        return items.convert_to_items(response.json())

    def get_stream(self, page_cursor=None):
        params = self._create_params()
        if page_cursor is not None:
            params['cursor'] = page_cursor
            pass

        response = self._request(self._create_url('me/activities/tracks/affiliated'),
                                 headers={'Accept': 'application/json'},
                                 params=params)
        self._handle_error(response)
        return items.convert_to_items(response.json())

    def follow_user(self, user_id, follow=True):
        method = 'PUT'
        if not follow:
            method = 'DELETE'
            pass

        response = self._request(self._create_url('me/followings/%s' % unicode(user_id)), method=method,
                                 headers={'Accept': 'application/json'})
        self._handle_error(response)
        return response.json()

    def like_track(self, track_id, like=True):
        method = 'PUT'
        if not like:
            method = 'DELETE'
            pass

        response = self._request(self._create_url('e1/me/track_likes/%s' % unicode(track_id)),
                                 method=method,
                                 headers={'Accept': 'application/json'})
        self._handle_error(response)
        return True

    def like_playlist(self, playlist_id, like=True):
        method = 'PUT'
        if not like:
            method = 'DELETE'
            pass

        response = self._request(self._create_url('e1/me/playlist_likes/%s' % unicode(playlist_id)),
                                 method=method,
                                 headers={'Accept': 'application/json'})
        self._handle_error(response)
        return True

    def _request(self, url, method='GET', headers=None, post_data=None, params=None, allow_redirects=True):
        if not params:
            params = {}
            pass

        if self._client_id:
            params['client_id'] = self._client_id
            pass

        if not headers:
            headers = {}
            pass

        if self._access_token:
            headers['Authorization'] = 'OAuth %s' % self._access_token
            pass

        return super(Client, self)._request(url, method, headers, post_data, params, allow_redirects)

    def login(self, username, password):
        post_data = {'grant_type': 'password',
                     'client_id': self._client_id,
                     'client_secret': self._client_secret,
                     'username': nightcrawler.utils.strings.to_utf8(username),
                     'password': nightcrawler.utils.strings.to_utf8(password),
                     'scope': 'non-expiring'}

        headers = {'Content-Type': 'multipart/form-data'}

        response = self._request(self._create_url('oauth2/token'), method='POST', headers=headers, post_data=post_data)
        self._handle_error(response)
        return response.json()

    pass


plugin.audio.soundcloud/resources/lib/nightcrawler/items.py
__author__ = 'bromix'

import hashlib

from . import utils


def create_item_hash(item):
    m = hashlib.md5()
    m.update(utils.strings.to_utf8(item['uri']))
    return m.hexdigest()


def create_next_page_item(context, thumbnail=None, fanart=None):
    if not fanart:
        fanart = context.get_fanart()
    page = int(context.get_param('page', 1)) + 1

    new_params = {}
    new_params.update(context.get_params())
    new_params['page'] = unicode(page)

    title = context.localize(30106)
    if title.find('%d') != -1:
        title %= page
        pass

    return {'type': 'folder',
            'title': title,
            'uri': context.create_uri(context.get_path(), new_params),
            'images': {'thumbnail': thumbnail,
                       'fanart': fanart}}


def create_search_item(context, thumbnail=None, fanart=None):
    if not thumbnail:
        thumbnail = context.create_resource_path('media/search.png')
    if not fanart:
        fanart = context.get_fanart()

    uri = context.create_uri('search/list')
    if context.get_search_history().get_max_item_count() == 0:
        uri = context.create_uri('search/query')

    return {'type': 'folder',
            'title': context.localize(30102),
            'uri': uri,
            'images': {'thumbnail': thumbnail,
                       'fanart': fanart}}


def create_watch_later_item(context, thumbnail=None, fanart=None):
    if not thumbnail:
        thumbnail = context.create_resource_path('media/watch_later.png')
    if not fanart:
        fanart = context.get_fanart()
    return {'type': 'folder',
            'title': context.localize(30107),
            'uri': context.create_uri('watch_later/list'),
            'images': {'thumbnail': thumbnail,
                       'fanart': fanart}}


plugin.audio.soundcloud/resources/lib/nightcrawler/provider.py
__author__ = 'bromix'

import json

from .core.nightcrawler_decorators import register_path, register_context_value, register_path_value
from .core.view_manager import ViewManager
from .exception import ProviderException


class Provider(object):
    LOCAL_SETUP_WIZARD_EXECUTE = 30030
    LOCAL_LOGIN_FAILED = 30121
    LOCAL_LOGIN_USERNAME = 30001
    LOCAL_LOGIN_PASSWORD = 30002

    LOCAL_PLEASE_WAIT = 30119

    LOCAL_FAVORITES = 30100
    LOCAL_FAVORITES_ADD = 30101
    LOCAL_FAVORITES_REMOVE = 30108

    LOCAL_WATCH_LATER = 30107
    LOCAL_WATCH_LATER_ADD = 30107
    LOCAL_WATCH_LATER_REMOVE = 30108

    LOCAL_SEARCH = 30102
    LOCAL_SEARCH_TITLE = 30102
    LOCAL_SEARCH_NEW = 30110
    LOCAL_SEARCH_RENAME = 30113
    LOCAL_SEARCH_REMOVE = 30108
    LOCAL_SEARCH_CLEAR = 30120

    LOCAL_CONFIRM_DELETE = 30114
    LOCAL_CONFIRM_REMOVE = 30115
    LOCAL_DELETE_CONTENT = 30116
    LOCAL_REMOVE_CONTENT = 30117

    LOCAL_SELECT_VIDEO_QUALITY = 30010

    LOCAL_SETUP_VIEW_DEFAULT = 30027
    LOCAL_SETUP_VIEW_VIDEOS = 30028

    LOCAL_LIBRARY = 30103
    LOCAL_HIGHLIGHTS = 30104
    LOCAL_ARCHIVE = 30105
    LOCAL_NEXT_PAGE = 30106

    LOCAL_LATEST_VIDEOS = 30109

    LOCAL_SETUP_OVERRIDE_VIEW = 30038

    PATH_SEARCH = '/search/list/'
    PATH_SEARCH_LIST = '/search/list/'
    PATH_SEARCH_QUERY = '/search/query/'
    PATH_SEARCH_CLEAR = '/search/clear/'
    PATH_SEARCH_RENAME = '/search/rename/'
    PATH_SEARCH_REMOVE = '/search/remove/'

    PATH_FAVORITES_ADD = '/favorites/add/'
    PATH_FAVORITES_LIST = '/favorites/list/'
    PATH_FAVORITES_REMOVE = '/favorites/remove/'

    PATH_WATCH_LATER_ADD = '/watch_later/add/'
    PATH_WATCH_LATER_LIST = '/watch_later/list/'
    PATH_WATCH_LATER_REMOVE = '/watch_later/remove/'

    def __init__(self):
        pass

    def _process_addon_setup(self, context):
        settings = context.get_settings()

        # exit if the setup isn't enabled
        if not settings.is_setup_wizard_enabled():
            return

        # exit if the setup shouldn't be executed
        if not context.get_ui().on_yes_no_input(context.get_name(), context.localize(self.LOCAL_SETUP_WIZARD_EXECUTE)):
            return

        view_manager = ViewManager(context, self)
        view_manager.setup()

        self.on_setup(context, mode='setup')

        # disable the setup
        settings.set_bool(settings.ADDON_SETUP, False)
        pass

    def on_setup(self, context, mode):
        if mode == 'content-types':
            return ['default']

        return None

    def select_video_stream(self, context, video_streams, video_quality_index=None, video_item=None):
        """
        Returns a selected video stream or False if the user aborted
        :param context: the current context
        :param video_streams: a list of selectable video streams
        :param video_item: (optional) video item to update with the uri of the selected video stream
        :param video_quality_index: index mapping to video quality
        :return:
        """
        if not video_quality_index:
            video_quality_index = [360, 720]
            pass

        def _sort_video_streams(_video_stream):
            return _video_stream.get('sort', 0)

        def _find_best_fit(_video_streams):
            _video_quality = context.get_settings().get_video_quality(video_quality_index=video_quality_index)
            _last_delta = None

            _selected_video_stream = None
            for _video_stream in _video_streams:
                _delta = abs(_video_stream.get('video', {}).get('height', 0) - _video_quality)
                if _selected_video_stream is None or _last_delta is None or _last_delta > _delta:
                    _last_delta = _delta
                    _selected_video_stream = _video_stream
                    pass
                pass

            return _selected_video_stream

        # sort stream, highest values first based on 'sort' array
        video_streams = sorted(video_streams, key=_sort_video_streams, reverse=True)

        # log all possible steams
        context.log_debug('selectable streams: %d' % len(video_streams))
        for video_stream in video_streams:
            context.log_debug('selectable stream: %s' % video_stream)
            pass

        # show a list of selectable streams (if enabled)
        selected_video_stream = None
        if context.get_settings().ask_for_video_quality() and len(video_streams) > 1:
            items = map(lambda x: (x['title'], x), video_streams)
            selected_video_stream = context.get_ui().on_select(context.localize(self.LOCAL_SELECT_VIDEO_QUALITY), items,
                                                               default=None)
            pass
        else:
            # fallback - use best fit
            selected_video_stream = _find_best_fit(video_streams)
            pass

        if selected_video_stream is None:
            return False

        # log selected video stream
        context.log_debug('selected stream: %s' % selected_video_stream)

        # we need an uri in the video stream
        if not 'uri' in selected_video_stream:
            raise ProviderException('Missing uri in video stream')

        # update the given video item (optional)
        if video_item:
            video_item['uri'] = selected_video_stream['uri']
            return video_item

        # return an uri item
        return {'type': 'uri',
                'uri': selected_video_stream['uri']}

    def navigate(self, context):
        # TODO - fix auth
        # self._process_addon_setup(context)

        method_names = dir(self)
        for method_name in method_names:
            method = getattr(self, method_name)
            if hasattr(method, 'nightcrawler_registered_path'):
                result = method(context)
                if result is not None:
                    return result

        raise ProviderException('Missing method for path "%s"' % context.get_path())

    def get_fanart(self, context):
        """
        Can be overriden by the derived class to return an alternate image
        :param context: the current image
        :return: full path to the fanart image
        """
        return context.get_fanart()

    def on_search(self, context, search_text):
        """
        The derived class has to implement this method in case of support for search
        :param context: the current context
        :param search_text: the search term in unicode
        :return: a list of items or False if something went wrong
        """
        raise NotImplementedError()

    @register_path('/favorites/(?P<method>add|remove)/')
    @register_path_value('method', unicode)
    @register_context_value('item', dict, required=True)
    def _internal_favorites_with_item(self, context, method, item):
        if method == 'add':
            context.get_favorite_list().add(item)
            return True

        if method == 'remove':
            context.get_favorite_list().remove(item)
            context.get_ui().refresh_container()
            return True

        return False

    @register_path('/favorites/list/')
    def on_favorites_list(self, context):
        result = context.get_favorite_list().list()
        for directory_item in result:
            context_menu = [(context.localize(self.LOCAL_WATCH_LATER_REMOVE),
                             'RunPlugin(%s)' % context.create_uri(self.PATH_FAVORITES_REMOVE,
                                                                  {'item': json.dumps(directory_item)}))]
            directory_item['context-menu'] = {'items': context_menu}
            pass

        return result

    @register_path('/watch_later/(?P<method>add|remove)/')
    @register_path_value('method', unicode)
    @register_context_value('item', dict, required=True)
    def _internal_watch_later_with_item(self, context, method, item):
        if method == 'add':
            context.get_watch_later_list().add(item)
            return True

        if method == 'remove':
            context.get_watch_later_list().remove(item)
            context.get_ui().refresh_container()
            return True

        return False

    @register_path('/watch_later/list/')
    def on_watch_later(self, context):
        video_items = context.get_watch_later_list().list()

        for video_item in video_items:
            context_menu = [(context.localize(self.LOCAL_WATCH_LATER_REMOVE),
                             'RunPlugin(%s)' % context.create_uri(self.PATH_WATCH_LATER_REMOVE,
                                                                  {'item': json.dumps(video_item)}))]
            video_item['context-menu'] = {'items': context_menu}
            pass

        return video_items

    @register_path('/search/(?P<method>(remove|rename))/')
    @register_path_value('method', unicode)
    @register_context_value('q', unicode, alias='query', required=True)
    def _internal_search_with_query(self, context, method, query):
        search_history = context.get_search_history()
        if method == 'remove':
            search_history.remove(query)
            context.get_ui().refresh_container()
            return True

        if method == 'rename':
            result, new_query = context.get_ui().on_keyboard_input(context.localize(self.LOCAL_SEARCH_RENAME), query)
            if result:
                search_history.rename(query, new_query)
                context.get_ui().refresh_container()
            return True

        return False

    @register_path('/search/(?P<method>(list|query|clear))/')
    @register_path_value('method', unicode)
    @register_context_value('q', unicode, alias='query', default=u'')
    def _internal_search_without_query(self, context, method, query):
        search_history = context.get_search_history()
        if method == 'clear':
            search_history.clear()
            context.get_ui().refresh_container()
            return True

        if method == 'list':
            # add new search
            result = [{'type': 'folder',
                       'title': '[B]%s[/B]' % context.localize(self.LOCAL_SEARCH_NEW),
                       'uri': context.create_uri(self.PATH_SEARCH_QUERY),
                       'images': {'thumbnail': context.create_resource_path('media/new_search.png'),
                                  'fanart': self.get_fanart(context)}}]

            for query in search_history.list():
                # we create a new instance of the SearchItem
                context_menu = [(context.localize(self.LOCAL_SEARCH_REMOVE),
                                 'RunPlugin(%s)' % context.create_uri(self.PATH_SEARCH_REMOVE, {'q': query})),
                                (context.localize(self.LOCAL_SEARCH_RENAME),
                                 'RunPlugin(%s)' % context.create_uri(self.PATH_SEARCH_RENAME, {'q': query})),
                                (context.localize(self.LOCAL_SEARCH_CLEAR),
                                 'RunPlugin(%s)' % context.create_uri(self.PATH_SEARCH_CLEAR))]
                item = {'type': 'folder',
                        'title': query,
                        'uri': context.create_uri(self.PATH_SEARCH_QUERY, {'q': query}),
                        'images': {'thumbnail': context.create_resource_path('media/search.png'),
                                   'fanart': self.get_fanart(context)},
                        'context-menu': {'items': context_menu}}
                result.append(item)

            return result

        if method == 'query':
            result = True
            if not query:
                result, query = context.get_ui().on_keyboard_input(context.localize(self.LOCAL_SEARCH_TITLE))

            if result and query:
                search_history.update(query)

                return self.on_search(context, query)

        return False

    def handle_exception(self, context, exception_to_handle):
        """
        Can be overridden by the derived class to handle exceptions
        :param context: the current context
        :param exception_to_handle: the caught exception
        :return: None if nothing can be done. True exception can be handled, False if not. A list of items is also possible.
        """
        return None

    def tear_down(self, context):
        """
        Can be overridden by the derived class to free resources or write some stuff to disk/cache
        :param context: the current context
        """
        pass
